### PR TITLE
Upgrade to ESLint v7

### PR DESCRIPTION
**Pull request by kaicataldo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            |
| Patch: Bug Fix?          | :+1:
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This updates all of our packages to have a peer dependency on ESLint v7 (the current release line), as well as to use v7 to lint the repo. The biggest difference is some stricter validation in the RuleTester (which is why I needed to update tests).

Opening this in favor of #11778.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/11857"></a>

### CODE REVIEW

**Comment by kaicataldo** [comment]
Removing `babel-eslint` from the top-level package.json leads to installation failing (locally as well).

>  error An unexpected error occurred: "ENOENT: no such file or directory, lstat '/home/runner/work/babel/babel/eslint/babel-

I think it's because `babel-eslint` has `semver` as a dependency and so the linked package is able to resolve the module. Is there another way we can consume the `babel-eslint-*` packages in the top-level `package.json` that will correctly install the deps?

**Comment by kaicataldo** [comment]
I _think_ I've found a way to use the local versions of the `babel-eslint-*` packages while still linking them in dev. Please let me know what you think!

File: scripts/integration-tests/e2e-babel-old-version.sh [hunk 1]
```diff
@@ -21,6 +21,10 @@ cd ../..
 #==============================================================================#
 
 startLocalRegistry "$PWD"/scripts/integration-tests/verdaccio-config.yml
+
+# Install dependencies in individual packages so that we can link them at the top level.
+for package in eslint/*/; do yarn --production --cwd $$package; done
```
Comment (kaicataldo) [inline scripts/integration-tests/e2e-babel-old-version.sh:26]: Alternative we could make this a Make target

**Comment by kaicataldo** [comment]
CI is currently failing because ESLint v7 has dropped support for Node <10 . This means that installation is failing on Node 6/8. I think it makes the most sense to use Yarn's `--ignore-engines` flag in Node 6/8. Thoughts?

**Comment by kaicataldo** [comment]
Actually, it looks like we already do this: https://github.com/babel/babel/blob/main/Makefile#L265. Hmmm...

**Comment by kaicataldo** [comment]
Oh I see, it's from the new Yarn installs I added 😄. Adding the flag now.

File: Makefile [hunk 2]
```diff
@@ -258,7 +258,7 @@ bootstrap-only: lerna-bootstrap
 
 yarn-install: clean-all
 	# Install dependencies in individual packages so that we can link them at the top level.
-	for package in eslint/*/; do yarn --production --cwd $$package; done
+	for package in eslint/*/; do yarn --cwd $$package; done
```
Comment (hzoo) [inline Makefile]: what's the difference between $$package and $package? (just looking at the other files)

Comment (JLHwung) [reply Makefile]: I think this is how `$` should be escaped in Makefile.

Comment (kaicataldo) [reply Makefile]: That's right! I had to look this up the first time I edited this file haha

**Pull request merged**